File: leet_lint_code/lint_code/majority_element_ii.cpp
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of lint code problem 47.

#ifndef LEETLINTCODE_LINTCODE_MAJORITYELEMENTII_CPP
#define LEETLINTCODE_LINTCODE_MAJORITYELEMENTII_CPP

#include <vector>
#include <iostream>

class SD {
 public:
  SD() {}
  SD(int value) {
    this->value = value;
    this->cnt = 1;
  }

  int value;
  int cnt;
};

class MajorityElementII {
 public:
  int majorityNumber(std::vector<int> &nums) {
    int len = (int)nums.size();
    if (2 >= len) return -1;

    std::vector<SD> v;

    int i = 1;
    v.push_back(SD(nums[0]));
    int sub_cnt = 0;

    while (i < len) {
      SD* tmp = find(v, nums[i]);
      if (NULL != tmp) {
        ++ tmp->cnt;
      } else if (v.size() < 3) {
        v.push_back(SD(nums[i]));
      } else {
        ++ sub_cnt;
        for (int j = 0; j < 3; ++ j) {
          -- v[j].cnt;
          if (v[j].cnt == 0) {
            v[j] = SD(nums[i]);
            for (size_t k = 0; k < j;  ++ k) {
              v[k].cnt += sub_cnt;
            }
            sub_cnt = 0;
          }
        }
      }
      ++ i;
    }

    int majority = v[0].value;
    int tmp = v[0].cnt;
    for (int i = 1; i < 3; ++ i) {
      if (tmp < v[i].cnt) {
        tmp = v[i].cnt;
        majority = v[i].value;
      }
    }

    return majority;
  }

  SD* find( std::vector<SD> &v, int value) {
    int len = (int)v.size();
    for (int i = 0; i < len; ++ i) {
      if (value == v[i].value) return &v[i];
    }
    return NULL;
  }
};

#endif // LEETLINTCODE_LINTCODE_MAJORITYELEMENTII_CPP

File: leet_lint_code/leet_code/java/MajorityElementII.java
// Copyright © 2018 Hailin Gu. All rights reserved.
// License
// Author: Hailin Gu
// This is a answer of leet code problem 229.
// Date: 2019.8.12


import java.util.ArrayList;
import java.util.List;

public class MajorityElementII {
    public List<Integer> majorityElement(int[] nums) {
        List<Integer> ans = new ArrayList<>();

        int ans1 = 0;
        int cnt1 = 0;

        int ans2 = 0;
        int cnt2 = 0;

        for (int num : nums) {
            if (cnt1 == 0 && num != ans2) {
                ans1 = num;
                cnt1++;
            } else if (ans1 == num) {
                ++cnt1;
            } else if (cnt2 == 0){
                ans2 = num;
                cnt2 ++;
            }else if (ans2 == num) {
                ++cnt2;
            } else {
                cnt1--;
                cnt2--;
            }
        }

        cnt1 = 0;
        cnt2 = 0;
        for (int num : nums) {
            if (num == ans1) {
                ++cnt1;
            } else if (num == ans2) {
                ++cnt2;
            }
        }

        if (cnt1 > nums.length / 3) {
            ans.add(ans1);
        }

        if (cnt2 > nums.length / 3) {
            ans.add(ans2);
        }

        return ans;
    }
}
